spec/integration/application/agent_spec.rb

```diff
@@ -0,0 +1,233 @@
+require 'puppet_spec/compiler'
+
+describe "puppet agent", unless: Puppet::Util::Platform.jruby? do
+  include PuppetSpec::Files
+  include PuppetSpec::Compiler
+
+  before :all do
+    WebMock.disable!
+  end
+
+  after :all do
+    WebMock.enable!
+  end
+
+  before :each do
+    # make sure we don't take too long
+    Puppet[:http_connect_timeout] = '5s'
+    Puppet[:server] = '127.0.0.1'
+    Puppet[:certname] = '127.0.0.1'
+
+    Puppet[:localcacert] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', 'ca.pem')
+    Puppet[:hostcrl] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', 'crl.pem')
+    Puppet[:hostprivkey] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', '127.0.0.1-key.pem')
+    Puppet[:hostcert] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', '127.0.0.1.pem')
+
+    facts = Puppet::Node::Facts.new(Puppet[:certname])
+    Puppet::Node::Facts.indirection.save(facts)
+  end
+
+  let(:server) { PuppetSpec::Puppetserver.new }
+  let(:agent) { Puppet::Application[:agent] }
+  let(:logfile) { tmpfile('agent.log') }
+  let(:args) { ['--test', '--logdest', logfile] }
+  let(:node) { Puppet::Node.new(Puppet[:certname], environment: 'production')}
+  let(:formatter) { Puppet::Network::FormatHandler.format(:rich_data_json) }
+
+  context 'server_list' do
+    it "uses the first server in the list" do
+      Puppet[:server_list] = '127.0.0.1'
+
+      server.start_server do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+      end
+    end
+
+    it "falls back, recording the first viable server in the report" do
+      server.start_server do |port|
+        Puppet[:server_list] = "puppet.example.com,#{Puppet[:server]}"
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+
+        report = Puppet::Transaction::Report.convert_from(:yaml, File.read(Puppet[:lastrunreport]))
+        expect(report.master_used).to eq("127.0.0.1:#{port}")
+      end
+    end
+
+    it "doesn't write a report if no servers could be contacted" do
+      Puppet[:server_list] = "puppet.example.com"
+
+      expect {
+        agent.command_line.args = args
+        agent.run
+      }.to exit_with(1)
+
+      # I'd expect puppet to update the last run report even if the server_list was
+      # exhausted, but it doesn't work that way currently, see PUP-6708
+      expect(File).to_not be_exist(Puppet[:lastrunreport])
+    end
+
+    it "omits master_used when not using server_list" do
+      server.start_server do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+      end
+
+      report = Puppet::Transaction::Report.convert_from(:yaml, File.read(Puppet[:lastrunreport]))
+      expect(report.master_used).to be_nil
+    end
+  end
+
+  context 'rich data' do
+    it "applies deferred values" do
```

Comment:
This test is from puppetdb CI. Apparently the only puppet beaker test for deferred functions is in https://github.com/puppetlabs/puppet/blob/6.14.0/acceptance/tests/pluginsync/feature/pluginsync_should_sync_features.rb#L94-L104. I left that test as-is, because it verifies a bunch of other "features" are available after pluginsync, such as a custom type, custom function, etc.